cli/azd/pkg/project/framework_service.go

```diff
@@ -79,22 +78,25 @@ type FrameworkService interface {
 	Restore(
 		ctx context.Context,
 		serviceConfig *ServiceConfig,
-	) *async.TaskWithProgress[*ServiceRestoreResult, ServiceProgress]
+		showProgress ShowProgress,
+	) (ServiceRestoreResult, error)
 
 	// Builds the source for the framework service
 	Build(
 		ctx context.Context,
 		serviceConfig *ServiceConfig,
 		restoreOutput *ServiceRestoreResult,
-	) *async.TaskWithProgress[*ServiceBuildResult, ServiceProgress]
+		showProgress ShowProgress,
+	) (ServiceBuildResult, error)
 
 	// Packages the source suitable for deployment
 	// This may optionally perform a rebuild internally depending on the language/framework requirements
 	Package(
 		ctx context.Context,
 		serviceConfig *ServiceConfig,
 		buildOutput *ServiceBuildResult,
-	) *async.TaskWithProgress[*ServicePackageResult, ServiceProgress]
+		showProgress ShowProgress,
+	) (ServicePackageResult, error)
```

Comment:
Similar to other comments.  This contract is to support the business/backend functionality as it relates to ServiceTargets and framework services.  Directly coupling progress display callbacks on this contacts violates the single responsibility principle and moves us more towards a tightly coupled architecture that would prevent us from reusing these services/contracts for other scenarios in the future.

We should not be designing for the future but want to limit us painting ourselves into a corner that would be hard to get out of.

Reply:
On a similar note - if a components 4 levels down wants to start participating in progress reporting this model would cause them to take a param dependency on the `showProgress` function all the way down the call stack.

An example of this would be inside of one of our tools.  We want to post some progress message about an tool operation that could take some time.

This would be difficult to achieve in the current model without tightly coupling the contract of the tool to now support this `showProgress` parameter.

Reply:
> Similar to other comments. This contract is to support the business/backend functionality as it relates to ServiceTargets and framework services. Directly coupling progress display callbacks on this contacts violates the single responsibility principle and moves us more towards a tightly coupled architecture that would prevent us from reusing these services/contracts for other scenarios in the future.

Can you elaborate on how this is "tightly coupled"? The methods are dependent on an abstract function whose description is to "show a progress message". It doesn't provide an implementation. The coupling here is unchanged from previous, where we had `async.TaskWithProgress`. The coupling here is also unchanged when dependent on a "Publish(msg string)` function.

> An example of this would be inside of one of our tools. We want to post some progress message about an tool operation that could take some time.

That isn't the scope of the changes or the problem as I understood it. We're trying to move away from async, and the first step is to move async -> sync. If we've decided that we want to have a UX pattern around the idea of:

- A task (spinner)
- A step (spinner title details)

We can certainly move in that direction, but I would not recommend doing it in a single change. Furthermore, this is one that I'd want us to solidify on the actual UX requirement here before codifying and implementing a design.